File: java/src/main/java/com/penapereira/patterns/chainofresponsibility/Handler.java
package com.penapereira.patterns.chainofresponsibility;

/** Stores the successor and implements the default: forward, or answer "unhandled". */
public abstract class Handler {

    private Handler next;

    /** Links the successor and returns it, so chains read left to right. */
    public Handler setNext(Handler next) {
        this.next = next;
        return next;
    }

    public String handle(int request) {
        return next == null ? "unhandled" : next.handle(request);
    }
}

File: javascript/src/chain-of-responsibility/handler.js
// Chain of Responsibility: pass a request along linked handlers until one takes it.

/** Stores the successor and implements the default: forward, or answer "unhandled". */
export class Handler {
  #next;

  /** Links the successor and returns it, so chains read left to right. */
  setNext(handler) {
    this.#next = handler;
    return handler;
  }

  handle(request) {
    return this.#next === undefined ? "unhandled" : this.#next.handle(request);
  }
}
